### PR TITLE
Use ArrayPool for oversized allocations

### PR DESCRIPTION
**Pull request by benaadams**
*When Pipelines allocation is larger than the max pool size; use ArrayPool, rather than `new byte[]`*

Oversized allocations are useful in situations where you want to minimize syscalls; for example `IHttpSendFileFeature` in Kestrel its faster to read from a file directly into the Pipe (zero copy(ish)) with a larger buffer size (16kB+) than it is to read the smaller chunks the Pool allows, however if its allocating an array each time than this more than offsets that advantage.

Currently with the smaller block sizes it is more performant to read a file chunk first into a larger array pool buffer (lower syscalls), then copy that file into many smaller Pipe buffers. 

It makes more sense to just read into the Pipe directly and avoid the additional copy. 

It is unfortunate that this approach is currently slower; as you either have to increase file read calls using the smaller buffers, or allocate a new array for each read with a larger buffer. 

This PR makes direct file reads into the Pipe a non-problematic option.

/cc @davidfowl @Tratcher @stephentoub 

Resolves https://github.com/dotnet/corefx/issues/39640

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs [hunk 1]
```diff
@@ -59,20 +59,18 @@ public void SetOwnedMemory(byte[] arrayPoolBuffer)
...
             {
                 owner.Dispose();
             }
-            else if (_memoryOwner is byte[] array)
+            else
             {
-                ArrayPool<byte>.Shared.Return(array);
+                Debug.Assert(_memoryOwner is byte[]);
+
+                byte[] pooolArray = Unsafe.As<byte[]>(_memoryOwner);
```
Comment (davidfowl) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs]: nit: too many o's

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 2]
```diff
@@ -235,8 +235,8 @@ private BufferSegment AllocateSegment(int sizeHint)
             }
             else
             {
-                // We can't use the pool so allocate an array
-                newSegment.SetUnownedMemory(new byte[sizeHint]);
+                // We can't use the pool so use the array pool
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Nit: "the pool" => "the memory pool" (so as to distinguish it from "the array pool")?  (Same in other places.)

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 3]
```diff
@@ -235,8 +235,8 @@ private BufferSegment AllocateSegment(int sizeHint)
...
+                newSegment.SetOwnedMemory(ArrayPool<byte>.Shared.Rent(sizeHint));
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Why is this case different from the `_pool is null` case? i.e. why does that one use GetSegmentSize and this one doesn't (in which case the conditions could also be combined)?  Same question for the writer below.

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Yeah should be able to

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 2 above)
Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Removed the comment as combined the two arraypool accesses

File: src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs [hunk 4]
```diff
@@ -59,20 +59,16 @@ public void SetOwnedMemory(byte[] arrayPoolBuffer)
...
         {
             if (_memoryOwner is IMemoryOwner<byte> owner)
             {
                 owner.Dispose();
             }
-            else if (_memoryOwner is byte[] array)
+            else
             {
-                ArrayPool<byte>.Shared.Return(array);
+                byte[] poolArray = (byte[])_memoryOwner;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs]: Could `_memoryOwner` ever be anything else other than `IMemoryOwner<byte>` or `byte[]`?

Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs]: Not anymore.

**Comment by Tratcher** [comment]
This will make https://github.com/aspnet/AspNetCore/issues/12632 a lot harder.

**Comment by davidfowl** [comment]
Not really, we need need to expose an option to clear on return.

**Comment by Tratcher** [comment]
`ArrayPool<byte>.Shared` isn't usable if you always want cleared memory, you can't rely on everybody else to clear on return. You need an option to clear on Rent or you need a separate pool that's always cleared.

**Comment by davidfowl** [comment]
Sounds like a reason to do this https://github.com/dotnet/corefx/issues/25843

**Pull request merged**